### PR TITLE
only allow $eq for updateOne and deleteOne commands

### PR DESCRIPTION
**Pull request by Yuqi-Du**
Add a filterAnalyzer rule for checking non-$eq usage for table updateOne and deleteOne command

fixes: #1729 

**Checklist**
- [x] Changes manually tested
- [x] Automated Tests added/updated
- [x] Documentation added/updated
- [x] CLA Signed: [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/WhereCQLClauseAnalyzer.java [hunk 1]
```diff
@@ -70,6 +70,8 @@ AnalysisStrategy getStrategy(WhereCQLClauseAnalyzer analyzer) {
                     analyzer::checkNoFilters,
                     analyzer::checkAllColumnsExist,
                     analyzer::checkNonPrimaryKeyFilters,
+                    analyzer::checkOnlyEqUsage,
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/WhereCQLClauseAnalyzer.java:73]: Fine for now but in future we probably will want to pre-create these `List`s to reduce garbage generation.
(so no change proposed for this PR, just mention as future improvement)

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/WhereCQLClauseAnalyzer.java [hunk 2]
```diff
@@ -70,6 +70,8 @@ AnalysisStrategy getStrategy(WhereCQLClauseAnalyzer analyzer) {
...
+                    // TODO, checkOnlyEqUsage rule make checkNoInFilterUsage redundant.
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/WhereCQLClauseAnalyzer.java:74]: Not sure I understand this comment; is it obsolete, or should it be rewritten?

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/WhereCQLClauseAnalyzer.java [hunk 3]
```diff
@@ -239,6 +241,39 @@ private void checkFilteringOnComplexColumns(Map<CqlIdentifier, TableFilter> iden
...
+        identifierToFilter.entrySet().stream()
+            .filter(
+                entry -> {
+                  TableFilter tableFilter = entry.getValue();
+                  if (!(tableFilter instanceof NativeTypeTableFilter<?> nativeTypeTableFilter)) {
+                    return true;
+                  }
+                  return nativeTypeTableFilter.operator != NativeTypeTableFilter.Operator.EQ;
+                })
+            .map((Map.Entry::getKey))
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/WhereCQLClauseAnalyzer.java:263]: nit: extra parentheses (remove) :)

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/WhereCQLClauseAnalyzer.java [hunk 4]
```diff
@@ -239,6 +241,39 @@ private void checkFilteringOnComplexColumns(Map<CqlIdentifier, TableFilter> iden
...
+   * UpdateOne.
+   */
+  private void checkOnlyEqUsage(Map<CqlIdentifier, TableFilter> identifierToFilter) {
+
+    var nonEqFilterColumns =
+        identifierToFilter.entrySet().stream()
+            .filter(
+                entry -> {
+                  TableFilter tableFilter = entry.getValue();
+                  if (!(tableFilter instanceof NativeTypeTableFilter<?> nativeTypeTableFilter)) {
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/WhereCQLClauseAnalyzer.java:258]: Why is this needed? And should it really return `true` (included as failing filter) instead of `false` (ignored, I think)

File: src/main/resources/errors.yaml [hunk 5]
```diff
@@ -300,6 +300,16 @@ request-errors:
...
 
+  - scope: FILTER
+    code: UNSUPPORTED_FILTER_FOR_UPDATE_ONE_DELETE_ONE
+    title: updateOne and deleteOne commands do not support filter operation other than `$eq`
+    body: |-
+      The Filtering using the non `$eq` filter operations can select more than one row, and so cannot be used with the `updateOne` and `deleteOne` commands as they can only modify one row.
+      
+      The command used an invalid filter on the columns: ${unsupportedFilterColumns}.
+      
+      Resend the command only using `$eq` filter.
```
Comment (tatu-at-datastax) [inline src/main/resources/errors.yaml:311]: maybe: "filter" -> "filters" ? (since one can have more than one)
also maybe rephrase "command only using" -> "command that only has"

File: src/test/java/io/stargate/sgv2/jsonapi/fixtures/testdata/WhereAnalyzerTestData.java [hunk 6]
```diff
@@ -175,6 +175,15 @@ public WhereAnalyzerFixture assertExceptionOnDurationColumns(CqlIdentifier... co
       return assertExceptionContains(warning);
     }
 
+    public WhereAnalyzerFixture assertExceptionOnNonEqFilerForUpdateOneAndDeleteOne(
+        CqlIdentifier... columns) {
+      var identifiers = Arrays.stream(columns).sorted(CQL_IDENTIFIER_COMPARATOR).toList();
+      var warning =
```
Comment (tatu-at-datastax) [inline src/test/java/io/stargate/sgv2/jsonapi/fixtures/testdata/WhereAnalyzerTestData.java:181]: nit: this is error, not warning? Maybe name as such

File: src/main/resources/errors.yaml (hunk 5 above)
Comment (amorton) [reply src/main/resources/errors.yaml:311]: Resend the command using only `$eq` filters.

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/WhereCQLClauseAnalyzer.java (hunk 2 above)
Comment (amorton) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/WhereCQLClauseAnalyzer.java:74]: agree, remove the $in check if not needed any more and remove the exception

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/WhereCQLClauseAnalyzer.java [hunk 7]
```diff
@@ -239,6 +241,39 @@ private void checkFilteringOnComplexColumns(Map<CqlIdentifier, TableFilter> iden
...
   }
 
+  /**
+   * Check if any non $eq filter is used.
+   *
+   * <p>e.g. Table UpdateOne and DeleteOne, DATA API does not know the exact modified row count in
+   * DB, so it needs to restrict on the filters. Make sure only $eq is used for DeleteOne and
+   * UpdateOne.
+   */
+  private void checkOnlyEqUsage(Map<CqlIdentifier, TableFilter> identifierToFilter) {
```
Comment (amorton) [inline src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/WhereCQLClauseAnalyzer.java:251]: from discussions: we should not allow a sort() clause for updateOne and deleteOne for Tables because CQL does not allow sorting on these commands. Create a new ticket if we need to do the work.

**Review by amorton — CHANGES_REQUESTED** [review]
some changes needed

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/WhereCQLClauseAnalyzer.java [hunk 8]
```diff
@@ -239,6 +241,39 @@ private void checkFilteringOnComplexColumns(Map<CqlIdentifier, TableFilter> iden
     }
   }
 
+  /**
+   * Check if any non $eq filter is used.
+   *
+   * <p>e.g. Table UpdateOne and DeleteOne, DATA API does not know the exact modified row count in
```
Comment (vkarpov15) [inline src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/WhereCQLClauseAnalyzer.java:247]: Isn't it also possible for `$eq` to match multiple rows? For example, would `deleteOne({ isDeleted: false })` delete multiple documents if multiple documents have `isDeleted: false`? Potentially silly question, I just wanted to make sure. Otherwise this PR looks reasonable.

Comment (tatu-at-datastax) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/WhereCQLClauseAnalyzer.java:247]: Yes, if we did not require filters to cover the whole primary key, but we do (for deletes and updates), I think. So only-eq-filters is a necessary but not sufficient constraints -- filtering on all primary key columns is also required.

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/WhereCQLClauseAnalyzer.java (hunk 2 above)
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/WhereCQLClauseAnalyzer.java:74]: removed.

File: src/test/java/io/stargate/sgv2/jsonapi/fixtures/testdata/WhereAnalyzerTestData.java (hunk 6 above)
Comment (Yuqi-Du) [reply src/test/java/io/stargate/sgv2/jsonapi/fixtures/testdata/WhereAnalyzerTestData.java:181]: thanks, good catch.

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/WhereCQLClauseAnalyzer.java (hunk 8 above)
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/WhereCQLClauseAnalyzer.java:247]: Thanks folks
Correct, for both deleteOne and updateOne, there is a rule checking (checkFullPrimaryKey) to enforce the full primary key is constraint in filter clause.
And yes, so `only-eq-filters is a necessary but not sufficient constraints`

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/WhereCQLClauseAnalyzer.java (hunk 4 above)
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/WhereCQLClauseAnalyzer.java:258]: only nativeTypeTableFilter has $eq, so if a tableFilter is not a nativeTypeTableFilter, then it is using operator other than $eq, so return true to keep that in the filter stream.

File: src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/WhereCQLClauseAnalyzer.java (hunk 7 above)
Comment (Yuqi-Du) [reply src/main/java/io/stargate/sgv2/jsonapi/service/operation/tables/WhereCQLClauseAnalyzer.java:251]: confirmed, these sort() on deleteOne and updateOne have already been covered by

- UNSUPPORTED_SORT_FOR_TABLE_UPDATE_COMMAND
- UNSUPPORTED_SORT_FOR_TABLE_DELETE_COMMAND

**Review by amorton — APPROVED** [review]
